### PR TITLE
Add a hook for when overlays are created

### PR DESCRIPTION
**Pull request by langston-barrett**
I'm going to use this to enter what Spacemacs calls a "transient state", rather than using the keymap (which is confusing when combined with Spacemacs/Evil modal keybindings).  

I don't think we actually need to require `cursor-sensor` for this to work? Can we just add the property anyway?

### CODE REVIEW

File: symbol-overlay.el [hunk 1]
```diff
@@ -156,6 +156,12 @@
   :group 'symbol-overlay
   :type 'float)
 
+(defcustom symbol-overlay-cursor-hook
+  '(lambda (window oldpos entered-or-left) '())
```
Comment (purcell) [inline symbol-overlay.el]: `#'ignore` would be a better default here.

File: symbol-overlay.el [hunk 2]
```diff
@@ -156,6 +156,12 @@
   :group 'symbol-overlay
   :type 'float)
 
+(defcustom symbol-overlay-cursor-hook
```
Comment (purcell) [inline symbol-overlay.el]: This isn't a hook: hooks are lists of functions. This is just a function, so it should be called something like `symbol-overlay-cursor-sensor-function`.

File: symbol-overlay.el [hunk 3]
```diff
@@ -156,6 +156,12 @@
...
+  '(lambda (window oldpos entered-or-left) '())
+  "A function that is put in the cursor-sensor-functions on every overlay."
```
Comment (purcell) [inline symbol-overlay.el]: When referring to an elisp symbol in the docstring, you should quote it with a backtick and an apostrophe.

File: symbol-overlay.el [hunk 4]
```diff
@@ -336,7 +342,10 @@ Otherwise apply `symbol-overlay-default-face'."
     (if face (progn (overlay-put ov 'face face)
                     (overlay-put ov 'keymap symbol-overlay-map)
                     (overlay-put ov 'evaporate t)
-                    (overlay-put ov 'symbol symbol))
+                    (overlay-put ov 'symbol symbol)
+                    (overlay-put ov
+                                 'cursor-sensor-functions
+                                 `(,symbol-overlay-cursor-hook)))
```
Comment (purcell) [inline symbol-overlay.el]: It's a bit of a red flag that this is setting an array value using a hard-coded list of one value. What if there were existing `cursor-sensor-functions` on the underlying text?

**Comment by siddharthist** [comment]
Thanks for the quick review! I implemented your more general overlay-creation hook, and was able to recover my special case:
```elisp
    (defun my/cursor-function (window oldpos entered-or-left)
      (when (equal entered-or-left 'entered)
        (message "entered")))
    (defun my/symbol-overlay-overlay-created-function (ov)
      (overlay-put ov 'cursor-sensor-functions '(my/cursor-function)))

    (add-to-list 'symbol-overlay-overlay-created-functions
                 'my/symbol-overlay-overlay-created-function)
```
Your detailed feedback was educational, I'm obviously a novice elisp programmer!

**Comment by langston-barrett** [comment]
@purcell This is ready for a second round of review

**Pull request merged**